Support for printing help from scala compiler without passing any sources

Gedochao opened the pull request:
Fixes #753

Gedochao committed:
Run scalac -help without the need for inputs for relevant commands

Gedochao committed:
Refactor SimpleScalaCompiler
  - extract duplicated runScalacLike code

Gedochao committed:
Add support for more scalac print commands

Gedochao committed:
Enable for scalac pure prefixes to be used without explicitly calling the --scalac-option param

Gedochao committed:
Add `--scalac-help` option as an alias for `--scalac-option -help`

Gedochao committed:
Add integration tests for scalac help & print options

Gedochao committed:
NIT Refactor
- fill in missing explicit type annotations
- remove unnecessary brackets

Gedochao committed:
Fill in missing scaladocs
- SimpleJavaCompiler
- SimpleScalaCompiler
- ScalacLikeCommand

Gedochao committed:
Add docs for printing compiler help

Gedochao committed:
Move `--scalac-help` to shared options root rather than a separate class

alexarchambault commented on modules/cli-options/src/main/scala/scala/cli/commands/ScalacOptions.scala:
```diff
@@ -17,7 +17,6 @@ final case class ScalacOptions(
     scalacOption: List[String] = Nil
 )
 // format: on
-
```
NIT Why remove this line?

Gedochao replied:
accidental, I reversed this change

alexarchambault commented on modules/cli-options/src/main/scala/scala/cli/commands/ScalacOptions.scala:
```diff
@@ -33,29 +32,34 @@ object ScalacOptions {
         args match {
-          case h :: t
-              if scalacOptionsPrefixes.exists(h.startsWith) &&
-              !scalacOptionsPurePrefixes.contains(h) =>
```
This shouldn't be removed. This prevents us from using `-V` / `-W` / `-X` / `-Y` ourselves (like `-X` as an alias for `--cross`)

alexarchambault replied:
Or we can keep it, and remove the current use of `-X` in the code base (look for `"X"`)

Gedochao replied:
ended up removing the current use of `-X`

alexarchambault commented on modules/cli/src/main/scala/scala/cli/commands/ScalacLikeCommand.scala:
```diff
@@ -0,0 +1,52 @@
+  * @tparam T
+  *   command options type
+  */
+trait ScalacLikeCommand[T] { self: ScalaCommand[T] =>
```
Instead of creating this trait, couldn't `maybePrintSimpleScalacOutput` be added directly to `ScalaCommand`, and have a signature like
```scala
def maybePrintSimpleScalacOutput(buildOptions: BuildOptions): Unit
```

alexarchambault commented on modules/cli/src/main/scala/scala/cli/commands/ScalacLikeCommand.scala:
```diff
@@ -0,0 +1,52 @@
+        if (shared.scalacHelp && !scalacOptions.contains("-help"))
+          scalacOptions.appended("-help")
+        else scalacOptions
+      if (updatedScalacOptions intersect ScalacOptions.ScalacPrintOptions.toSeq).nonEmpty
```
```suggestion
      if updatedScalacOptions.exists(ScalacOptions.ScalacPrintOptions)
```
should be more efficient (iterates once on `updatedScalacOptions` and stops as soon as it finds a print option, rather than computing an intersection, which is more costly)

alexarchambault commented on modules/cli-options/src/main/scala/scala/cli/commands/ScalacOptions.scala:
```diff
@@ -33,29 +32,34 @@ object ScalacOptions {
     new Argument[List[String]] {
 
-      val underlying = StandardArgument[List[String]](scalacOptionsArg)
+      val underlying: StandardArgument[List[String]] = StandardArgument(scalacOptionsArg)
```
Do we really need to add that type ascription? The type was pretty clear (it appears explicitly) in the former code.

Gedochao replied:
<img width="620" alt="image" src="https://user-images.githubusercontent.com/18601388/171147060-0a10adeb-b68a-4890-8209-9015257d3812.png">

We don't really need to, but IntelliJ warns you about it by default.
IMO the codebase is more welcoming for new contributors if we keep the yellow highlights in their IDEs to the bare minimum.

alexarchambault replied:
That warning looks spurious to me, that field isn't public (it's an anonymous class, so it doesn't appear in the public API of this module).

Gedochao replied:
I don't think IntelliJ is smart enough to notice the difference 🧠
Still, doesn't hurt us to keep the type explicit here, I think

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/RunTestDefinitions.scala:
```diff
@@ -1698,4 +1699,51 @@ abstract class RunTestDefinitions(val scalaVersionOpt: Option[String])
+        "-help"
+      )
+        .call(cwd = root, mergeErrIntoOut = true)
+      expect(res2.exitCode == 0)
```
No need to check this (`.call` would fail if that wasn't the case)

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/RunTestDefinitions.scala:
```diff
@@ -1698,4 +1699,51 @@ abstract class RunTestDefinitions(val scalaVersionOpt: Option[String])
+          printOption
+        )
+          .call(cwd = root, mergeErrIntoOut = true)
+        expect(res.exitCode == 0)
```
Same as above (unnecessary)

Gedochao committed:
Apply suggestions from code review
- move `maybePrintSimpleScalacOutput` back to `ScalaCommand`
- remove ScalacLikeCommand trait
- remove the `-X` alias for cross-compilation
- remove unnecessary exit status code verification in Run command ITs
